Wip gc get source and dest zones

AlvinYangZF opened the pull request:
Merge @shampoo365 ‘s commit into this branch and commit to master together
@soumendus this pull requset fixed some of the issues that your comments mentioned.

skyzh committed:
fs: only acquire lock once on deletion

skyzh committed:
util: add dump metadata command

This PR adds dump metadata command. This command will dump all
files and extents in json format.

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
util: add metadata analysis tool

This commit adds metadata analysis tool. Refer to README.md
in util directory for more information.

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh commented on fs/io_zenfs.cc:
```diff
@@ -659,6 +660,64 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
+    IOStatus s;
+    s = zone_idx->Reset();
+    if (!s.ok()) {
+      // Debug(logger_, "Failed resetting zone when executing GC!");
```
Why comment this out? I believe it is needed to report an error if reset failed.

AlvinYangZF replied:
Added it back

skyzh commented on fs/io_zenfs.cc:
```diff
@@ -659,6 +660,64 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
+std::vector<Zone*> ZenFSGCWorker::GetDestZoneToMoveValidData(
+    uint64_t ttl_residue) {
+  while (ttl_residue) {
+    Zone* zone = zbd_->AllocateZone(Env::WLTH_NOT_SET);
```
Do we have any consideration on zone lifetime here? If leave this with not set, new extents with different lifetime will be pushed onto this zone, making it (possibly) harder to collect garbage.

skyzh replied:
And what if zone allocation fails?

AlvinYangZF commented on fs/zbd_zenfs.cc:
```diff
@@ -174,6 +174,19 @@ Zone *ZonedBlockDevice::GetIOZone(uint64_t offset) {
+  for (const auto z : io_zones) {
+    if ((!z->IsUsed()) && (!z->IsFull())) continue;
+    /* valid data in a zone is larger than 10% max capacity */
+    if ((z->max_capacity_ / z->used_capacity_ >= 10)) {
```
The 10% valid data in a zone should be described as z->used_capacity_ / z->max_capacity, but I didn't want to use floating varables, so I exchanged them. However, it is not readable and the denominator could be 0, so I think it is better to correct it.

skyzh replied:
You can use something like `0.1 * used_capacity_`, and I believe the compiler is smart enough to optimize it into integer multiplication.

skyzh commented on fs/zbd_zenfs.cc:
```diff
@@ -174,6 +174,20 @@ Zone *ZonedBlockDevice::GetIOZone(uint64_t offset) {
+    if ((!z->IsUsed()) && (!z->IsFull())) continue;
+    /* once the valid data in a zone is less than 10%, it will be seleted as GC
+     * zone */
+    if ((z->used_capacity_ * ZENFS_GC_THREASHOLD) / z->max_capacity_ <= 1) {
```
From my perspective, GC is not called very frequently, and we might need to dynamically configure the threshold. So I would prefer to simply use double multiplication.

skyzh reviewed (commented):
Rest LGTM. There might be some concurrency issues inside, but the current codebase is not complete enough to conduct a full review. I will find and review these issues later.

skyzh reviewed (approved):
LGTM for this part. Thanks for your work.

AlvinYangZF committed:
add gc worker
add header to fix compiling issue
get valid zone for GC source zones
fix gc threshold calculation, destination zone allocation and debugging message
get source zone's lifetime
add zones, extents and files to merge_zone_list, extent_list and  files_moved_to_dst_zone
change the valid rate to double
